lib/stellar/transaction_builder.rb

```diff
@@ -0,0 +1,108 @@
+module Stellar
+  class TransactionBuilder
+
+    attr_reader :source_account, :sequence_number, :base_fee, :time_bounds, :time_bounds, :memo, :operations
+
+    def initialize(
+      source_account:, 
+      sequence_number:, 
+      base_fee: 100, 
+      time_bounds: nil, 
+      memo: nil
+    )
+      raise ArgumentError, "bad :source_account" unless source_account.is_a?(Stellar::KeyPair)
+      raise ArgumentError, "bad :sequence_number" unless sequence_number.is_a?(Integer) && sequence_number > 0
+      raise ArgumentError, "bad :time_bounds" unless time_bounds.is_a?(Stellar::TimeBounds) || time_bounds.nil?
+      raise ArgumentError, "bad :base_fee" unless base_fee.is_a?(Integer) && base_fee >= 100
+
+      @source_account = source_account
+      @sequence_number = sequence_number
+      @base_fee = base_fee
+      @time_bounds = time_bounds
+      @memo = self.make_memo(memo)
+      @operations = Array.new
+    end
+
+    def build
+      if @time_bounds.nil?
+        raise "TransactionBuilder.time_bounds must be set during initialization or by calling set_timeout"
+      elsif !@time_bounds.min_time.is_a?(Integer) or !@time_bounds.max_time.is_a?(Integer)
+        raise "TimeBounds.min_time and max_time must be Integers"
+      elsif @time_bounds.max_time != 0 and @time_bounds.min_time > @time_bounds.max_time
+        raise "Timebounds.max_time must be greater than min_time"
+      elsif @time_bounds.max_time != 0 and @time_bounds.max_time < Time.now.to_i
+        raise "Timebounds.max_time must be in the future"
+      end
+      tx = Stellar::Transaction.new(
+        source_account: @source_account.account_id,
+        fee: @base_fee * @operations.length,
+        seq_num: @sequence_number,
+        time_bounds: @time_bounds,
+        memo: @memo,
+        operations: @operations,
+        ext: Stellar::Transaction::Ext.new(0)
+      )
+      @sequence_number += 1
+      tx
+    end
+
+    def add_operation(operation)
+      raise ArgumentError, "bad operation" unless operation.is_a? Stellar::Operation
+      @operations.push(operation)
+      self
+    end
+
+    def clear_operations()
+      @operations.clear
+      self
+    end
+
+    def set_sequence_number(seq_num)
+      raise ArgumentError, "bad sequence number" unless seq_num.is_a?(Integer) and seq_num > 0
+      @sequence_number = seq_num
+      self
+    end
+
+    def set_timeout(timeout)
+      if !timeout.is_a?(Integer) or timeout < 0
+        raise ArgumentError, "timeout must be a non-negative integer"
+      end
+
+      if @time_bounds.nil?
+        @time_bounds = Stellar::TimeBounds.new(min_time: 0, max_time: nil)
+      end
```

Comment:
If I was calling `set_timeout` I'd expect it to replace the entire time_bounds. Why do we need to preserve the existing `min_time`?

As a bonus if you have this function always overwrite the entire `@time_bounds` the few lines below will also become simple since you can default `max_time` to zero and only need to set it if its not zero.

Reply:
The [JS SDK](https://stellar.github.io/js-stellar-sdk/node_modules_stellar-base_src_transaction_builder.js.html#line139) does create a new object but it preserves the existing `min_time`.

I would suspect its because the name of the function is `set_timeout`, not `set_time_bounds`, which sounds like it is only concerned with how long the transaction should attempt to be executed.

We could replace this with a `set_time_bounds` function but I personally think this is a better function because users don't have to construct a `TimeBounds` or mess with `Time` objects.

Reply:
Got it. 👍